cpp/src/arrow/compute/exec/union_node.cc

```diff
@@ -0,0 +1,150 @@
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <mutex>
+
+#include "arrow/api.h"
+#include "arrow/compute/api.h"
+#include "arrow/compute/exec/exec_plan.h"
+#include "arrow/compute/exec/options.h"
+#include "arrow/compute/exec/util.h"
+#include "arrow/util/bitmap_ops.h"
+#include "arrow/util/checked_cast.h"
+#include "arrow/util/future.h"
+#include "arrow/util/logging.h"
+#include "arrow/util/thread_pool.h"
+
+namespace arrow {
+
+using internal::checked_cast;
+
+namespace compute {
+
+namespace {
+std::vector<std::string> GetInputLabels(ExecNode::NodeVector inputs) {
+  std::vector<std::string> labels(inputs.size());
+  for (size_t i = 0; i < inputs.size(); i++) {
+    labels[i] = "input_" + std::to_string(i) + "_label";
+  }
+  return labels;
+}
+}  // namespace
+struct UnionNode : ExecNode {
+  UnionNode(ExecPlan* plan, std::vector<ExecNode*> inputs)
+      : ExecNode(plan, inputs, GetInputLabels(inputs),
+                 /*output_schema=*/inputs[0]->output_schema(),
+                 /*num_outputs=*/1) {
+    if (this->input_count_.SetTotal(static_cast<int>(inputs.size()))) {
+      finished_.MarkFinished();
+    }
+  }
+
+  const char* kind_name() override { return "UnionNode"; }
+
+  static Result<ExecNode*> Make(ExecPlan* plan, std::vector<ExecNode*> inputs,
+                                const ExecNodeOptions& options) {
+    RETURN_NOT_OK(ValidateExecNodeInputs(plan, inputs, inputs.size(), "UnionNode"));
+    if (inputs.size() < 2) {
+      return Status::Invalid("Constructing a `UnionNode` with inputs size less than 2");
+    }
+    auto schema = inputs.at(0)->output_schema();
+    for (auto input : inputs) {
+      if (!input->output_schema()->Equals(schema)) {
+        return Status::Invalid(
+            "Constructing a `UnionNode` with inputs with different schemas");
+      }
+    }
+    return plan->EmplaceNode<UnionNode>(plan, std::move(inputs));
+  }
+
+  inline bool IsLeftInput(ExecNode* input) { return input == inputs_[0]; }
+
+  void InputReceived(ExecNode* input, int seq, ExecBatch batch) override {
+    ARROW_DCHECK(std::find_if(inputs_.begin(), inputs_.end(), [input](ExecNode* n) {
+                   return n == input;
+                 }) != inputs_.end());
+
+    if (finished_.is_finished()) {
+      return;
+    }
+    outputs_[0]->InputReceived(this, seq, std::move(batch));
```

Comment:
I wonder if it actually makes sense to forward seq like this. As I understand nodes can't really rely on this anyways so I don't think it harms things, but it seems odd to even have the parameter in that case. CC @bkietz for opinions. 

(We could modify `Increment` to also give the count, if we wanted to give a unique seq_num to each batch we emit here.)

Reply:
I think `seq` should be ultimately removed. It's a confusing parameter which we don't provide guarantees on. I'll add a JIRA for this anon

Reply:
Thanks. @aocsa we can leave this as-is then.

Reply:
https://issues.apache.org/jira/browse/ARROW-13660